Fix build caching related error `LNK4020`

Occasionally builds fail with:
> error LNK4020: a type record in 'd:\a\ophd\ophd\nas2d-core\.build\release_x86_nas2d\NAS2D.pdb' is corrupted; some symbols and types may not be accessible from the debugger [D:\a\OPHD\OPHD\appOPHD\appOPHD.vcxproj]

This seems to happen when the NAS2D submodule is updated. It's not clear what the exact cause is, though clearing cache entries and re-running the failed workflows can get them to pass. The caches are found at:
- https://github.com/OutpostUniverse/OPHD/actions/caches

Using the filter feature, entries to clear are (note no space after `:`):
- `key:buildCache`
- `key:nas2dCache`
- `key:vcpkgCache`

It's possible only a subset of those need to be cleared out.

----

I suspect this may also be related to new `vcpkg` packages. When the error shows up, there also tend to be newly built `vcpkg` packages.

Perhaps what we really need is a way to invalidate build caches when new `vcpkg` packages are available. This might be detected by a version change with:
```sh
vcpkg --version
```

Perhaps the `vcpkg` version string could become part of the cache key.

----

On a possibly related note, maybe the `baseline` value is of interest:
- https://learn.microsoft.com/en-us/vcpkg/users/versioning#version-constraints

----

We might only care if dependencies we rely on are updated. That may be less often than the version of `vcpkg` is updated. Perhaps using the `vcpkg --version` value would cause more frequent cache invalidation than is actually necessary.

Adding link to a recent run that showed more `LNK4020` errors:
https://github.com/OutpostUniverse/OPHD/actions/runs/13331042983

The first attempt at the run failed, then caches were cleared for that branch, and each job was re-run again. The third attempt shows both runs completed after clearing the cache.

The caches restored for the first failed run were:
> Cache restored from key: `vcpkgCache-Windows-x64-00fc58e29020291fd374aafc403c8590fa625c3e181b59d3d363ff734f63cfdd`
> Cache restored from key: `nas2dCache-Windows-x64-bc0abae348b5c0d4aa3552437e894670b9fff5ae`
> Cache restored from key: `buildCache-Windows-x64-afcb1dfc8af8b52d4befccdee88f955ea5ccd346`

This resulted in `error LNK4020`.

For the 3rd successful re-run of the job:
> Cache restored from key: `vcpkgCache-Windows-x64-00fc58e29020291fd374aafc403c8590fa625c3e181b59d3d363ff734f63cfdd`
> Cache restored from key: `nas2dCache-Windows-x64-bc0abae348b5c0d4aa3552437e894670b9fff5ae`
> Cache restored from key: `buildCache-Windows-x64-ba0fa9b5c28bfdc9a2fdfe281bd4d963def4bf6a`

As can be seen, only the last cache for the OPHD incremental build cache was different. That cache came from an earlier branch build, before a forced push. For the 3rd re-run, all three caches would have come from `main`. The first two caches were identical to `main`.

There was an earlier successful build, before a forced push here:
https://github.com/OutpostUniverse/OPHD/actions/runs/13318018422/job/37196668458
This resulted in saving the cache:
> Cache saved with key: `buildCache-Windows-x64-afcb1dfc8af8b52d4befccdee88f955ea5ccd346`

This appears to be the cache that was restored by the failed build above.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -160,7 +160,7 @@ jobs:
 
     - name: Save incremental build cache - OPHD
       uses: actions/cache/save@v4
-      if: steps.cacheRestoreOphd.outputs.cache-hit != 'true' || steps.cacheRestoreOphd.outcome == 'skipped'
+      if: steps.cacheRestoreOphd.outputs.cache-hit != 'true'
       with:
         path: .build
         key: ${{ env.cacheKeyOphd }}

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -75,7 +75,7 @@ jobs:
       if: steps.cacheRestoreVcpkg.outputs.cache-hit != 'true'
       with:
         path: vcpkg_installed
-        key: ${{ steps.cacheRestoreVcpkg.outputs.cache-primary-key }}
+        key: ${{ env.cacheKeyVcpkg }}
 
     - name: Copy vcpkg cache to NAS2D folder
       run: |
@@ -114,7 +114,7 @@ jobs:
       if: steps.cacheRestoreNas2d.outputs.cache-hit != 'true'
       with:
         path: nas2d-core/.build/
-        key: ${{ steps.cacheRestoreNas2d.outputs.cache-primary-key }}
+        key: ${{ env.cacheKeyNas2d }}
 
     - name: Restore incremental build cache
       uses: actions/cache/restore@v4

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -46,10 +46,10 @@ jobs:
         ophdSha="${{ github.sha }}"
 
         config="${{ runner.os }}-${{ matrix.platform }}"
-        cacheKeyPrefixOphd="buildCache-${config}-"
+        cacheKeyPrefixOphd="ophd-${config}-"
 
-        cacheKeyVcpkg="vcpkgCache-${config}-${vcpkgSha::8}"
-        cacheKeyNas2d="nas2dCache-${config}-${vcpkgSha::8}-${nas2dSha::8}"
+        cacheKeyVcpkg="vcpkg-${config}-${vcpkgSha::8}"
+        cacheKeyNas2d="nas2d-${config}-${vcpkgSha::8}-${nas2dSha::8}"
         cacheKeyOphd="${cacheKeyPrefixOphd}-${vcpkgSha::8}-${nas2dSha::8}-${ophdSha::8}"
 
         echo "cacheKeyVcpkg=${cacheKeyVcpkg}" >> $GITHUB_ENV

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -119,7 +119,7 @@ jobs:
     - name: Restore incremental build cache
       uses: actions/cache/restore@v4
       id: cacheRestoreOphd
-      if: github.ref != format('refs/heads/{0}', github.event.repository.default_branch) && steps.cacheRestoreNas2d.outputs.cache-hit == 'true'
+      if: steps.cacheRestoreNas2d.outputs.cache-hit == 'true' && github.ref != format('refs/heads/{0}', github.event.repository.default_branch)
       with:
         path: .build
         key: ${{ env.cacheKeyOphd }}
@@ -160,7 +160,7 @@ jobs:
 
     - name: Save incremental build cache - OPHD
       uses: actions/cache/save@v4
-      if: steps.cacheRestoreOphd.outcome == 'skipped' || steps.cacheRestoreOphd.outputs.cache-hit != 'true'
+      if: steps.cacheRestoreOphd.outputs.cache-hit != 'true' || steps.cacheRestoreOphd.outcome == 'skipped'
       with:
         path: .build
         key: ${{ env.cacheKeyOphd }}

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -48,9 +48,9 @@ jobs:
         config="${{ runner.os }}-${{ matrix.platform }}"
         cacheKeyPrefixOphd="buildCache-${config}-"
 
-        cacheKeyVcpkg="vcpkgCache-${config}-${vcpkgSha}"
-        cacheKeyNas2d="nas2dCache-${config}-${nas2dSha}"
-        cacheKeyOphd="${cacheKeyPrefixOphd}-${ophdSha}"
+        cacheKeyVcpkg="vcpkgCache-${config}-${vcpkgSha::8}"
+        cacheKeyNas2d="nas2dCache-${config}-${vcpkgSha::8}-${nas2dSha::8}"
+        cacheKeyOphd="${cacheKeyPrefixOphd}-${vcpkgSha::8}-${nas2dSha::8}-${ophdSha::8}"
 
         echo "cacheKeyVcpkg=${cacheKeyVcpkg}" >> $GITHUB_ENV
         echo "cacheKeyNas2d=${cacheKeyNas2d}" >> $GITHUB_ENV

